connect/runtime/src/test/java/org/apache/kafka/connect/integration/ExactlyOnceSourceIntegrationTest.java

```diff
@@ -0,0 +1,997 @@
+                for (long i = priorBoundary + 1; i < nextBoundary + 1; i++) {
+                    expectedRecordSeqnos.remove(i);
+                }
+            }
+            nextBoundary += priorBoundary;
+            priorBoundary = nextBoundary - priorBoundary;
+        }
+        List<Long> actualRecordSeqnos = new ArrayList<>();
+        sourceRecords.forEach(record -> actualRecordSeqnos.add(parseAndAssertValueForSingleTask(record)));
+        Collections.sort(actualRecordSeqnos);
+        assertEquals("Committed records should exclude connector-aborted transactions",
+                expectedRecordSeqnos, actualRecordSeqnos.subList(0, expectedRecordSeqnos.size()));
+    }
+
+    /**
+     * Brings up a one-node cluster, then intentionally fences out the transactional producer used by the leader
+     * for writes to the config topic to simulate a zombie leader being active in the cluster. The leader should
+     * automatically recover, verify that it is still the leader, and then succeed to create a connector when the
+     * user resends  the request.
+     */
+    @Test
+    public void testFencedLeaderRecovery() throws Exception {
+        // Much slower offset commit interval; should never be triggered during this test
+        workerProps.put(WorkerConfig.OFFSET_COMMIT_INTERVAL_MS_CONFIG, "600000");
+        startConnect();
+
+        String topic = "test-topic";
+        connect.kafka().createTopic(topic, 3);
+
+        int recordsProduced = 100;
+
+        Map<String, String> props = new HashMap<>();
+        props.put(CONNECTOR_CLASS_CONFIG, MonitorableSourceConnector.class.getName());
+        props.put(TASKS_MAX_CONFIG, "1");
+        props.put(TOPIC_CONFIG, topic);
+        props.put(KEY_CONVERTER_CLASS_CONFIG, StringConverter.class.getName());
+        props.put(VALUE_CONVERTER_CLASS_CONFIG, StringConverter.class.getName());
+        props.put(NAME_CONFIG, CONNECTOR_NAME);
+        props.put(TRANSACTION_BOUNDARY_CONFIG, POLL.toString());
+        props.put("messages.per.poll", Integer.toString(recordsProduced));
+
+        // expect all records to be consumed and committed by the connector
+        connectorHandle.expectedRecords(recordsProduced);
+        connectorHandle.expectedCommits(recordsProduced);
+
+        // make sure the worker is actually up (otherwise, it may fence out our simulated zombie leader, instead of the other way around)
+        assertEquals(404, connect.requestGet(connect.endpointForResource("connectors/nonexistent")).getStatus());
+
+        // fence out the leader of the cluster
+        transactionalProducer(DistributedConfig.transactionalProducerId(CLUSTER_GROUP_ID)).initTransactions();
+
+        // start a source connector--should fail the first time
+        assertThrows(ConnectRestException.class, () -> connect.configureConnector(CONNECTOR_NAME, props));
+
+        // if at first you don't succeed, then spam the worker with rest requests until it gives in to your demands
```

Comment:
This doesn't really explain why this second call is necessary.

Reply:
Ack, updated to something less snarky.